### PR TITLE
Adding the engineer view

### PR DESCRIPTION
**Pull request by ptrnull**
Adding a new view that shows info per engineer:

- Adding a new function to get the user stats: generate_user_stats
- Adding a new parameter 'engineer' on generate_histogram_stats to get the histogram data filtered by user
- Adding a new parameter 'engineer' on get_new_comments
- Adding a new ui template to show the engineer data: engineer.html
- Modifying macros.html to get the users with a link to their views
- Modifying ui.py to have a new route defined

### CODE REVIEW

File: dashboard/src/t5gweb/ui.py [hunk 1]
```diff
@@ -410,3 +411,37 @@ def get_account(account):
...
         histogram_stats=histogram_stats,
     )
+
+@BP.route("/engineer/<string:engineer>")
+@login_required
+def get_engineer(engineer):
+    """show bugs, cases and stats by for a given engineer"""
+    load_data()
+    stats = generate_user_stats(engineer)
+    #comments = get_new_comments(new_comments_only=False, account=engineer)
```
Comment (adhil0) [inline dashboard/src/t5gweb/ui.py:421]: NIT: Remove?

File: dashboard/src/t5gweb/templates/ui/engineer.html [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                                        <td class="issue-middle text-center">{{ issue['status'] }}</td>
+                                    </tr>
+                                {% endfor %}
+                            {% endif %}
+                        {% endfor %}
+                    {% endfor %}
+                {% endfor %}
+            </tbody>
+        </table>
+    </div>-->
```
Comment (adhil0) [inline dashboard/src/t5gweb/templates/ui/engineer.html:239]: NIT: Remove?

**Review by adhil0 — CHANGES_REQUESTED** [review]
The changes look good in my dev environment, and I'll think this feature will be very useful for all of the engineers.

Our linting check is throwing a couple of errors, though:

1. There are a couple of lines which exceed 88 characters. This can be fixed by running `black` or `ruff format`. (Unless you you disagree with the 88 char limit, in which case we can have a discussion around that)
2. `generate_user_stats()` is quite similar to `generate_stats()`, could we combine the two into a single, generalized function? Just to reduce duplication.

NOTE: The duplicated code detector is being too aggressive (it's flagging [this](https://github.com/RHsyseng/t5g-field-support-team-utils/actions/runs/8204826428/job/22440292239#step:4:321) innocent code as a duplication).  https://github.com/RHsyseng/t5g-field-support-team-utils/issues/268

NOTE: This PR has made me realize that there is quite a bit of duplication across all of our jinja templates, I've opened an issue to address this https://github.com/RHsyseng/t5g-field-support-team-utils/issues/267

File: dashboard/src/t5gweb/libtelco5g.py [hunk 3]
```diff
@@ -549,6 +549,143 @@ def generate_stats(account=None):
 
     return stats
 
+def generate_user_stats(engineer=None):
```
Comment (dcritch) [inline dashboard/src/t5gweb/libtelco5g.py:552]: rather than have a new function, could you modify `generate_stats` to take engineer as optional? similar to how you updated `generate_histogram_stats` and `get_new_comments`

File: dashboard/src/t5gweb/templates/ui/engineer.html [hunk 4]
```diff
@@ -0,0 +1,322 @@
+{% extends "skeleton.html" %}
```
Comment (dcritch) [inline dashboard/src/t5gweb/templates/ui/engineer.html:1]: on a similar note, this duplicates a lot of `account.html`. you could add some conditionals to change the header based on whether `engineer` or `account` was supplied, and only show the jira/bug table if its by account. maybe rename account.html to something more generic like `details.html` or something like that?

File: dashboard/src/t5gweb/ui.py (hunk 1 above)
Comment (ptrnull) [reply dashboard/src/t5gweb/ui.py]: ack, removed

File: dashboard/src/t5gweb/templates/ui/engineer.html (hunk 2 above)
Comment (ptrnull) [reply dashboard/src/t5gweb/templates/ui/engineer.html]: merged engineer into account

File: dashboard/src/t5gweb/libtelco5g.py (hunk 3 above)
Comment (ptrnull) [reply dashboard/src/t5gweb/libtelco5g.py]: you are right, I did miss this one, thanks!

File: dashboard/src/t5gweb/templates/ui/engineer.html (hunk 4 above)
Comment (ptrnull) [reply dashboard/src/t5gweb/templates/ui/engineer.html]: agree, merged engineer.html template into account.html, no need for a new one, if needed in the future we can extend it and rename it to details.html

File: dashboard/src/t5gweb/libtelco5g.py [hunk 5]
```diff
@@ -425,7 +425,8 @@ def generate_stats(account=None):
     if account is not None:
         logging.warning("filtering cases for {}".format(account))
         cards = {c: d for (c, d) in cards.items() if d["account"] == account}
-        cases = {c: d for (c, d) in cases.items() if d["account"] == account}
+        if engineer is None:
+            cases = {c: d for (c, d) in cases.items() if d["account"] == account}
```
Comment (adhil0) [inline dashboard/src/t5gweb/libtelco5g.py:429]: This logic results in an incorrect "Open Case" count in the stats table. Instead of showing the number of cases that the engineer is working on, it displays the total number of cases. 

I've been trying this, but the numbers aren't adding up:
```python
    if account is not None:
        logging.warning("filtering cases for {}".format(account))
        cards = {c: d for (c, d) in cards.items() if d["account"] == account}
        cases = {c: d for (c, d) in cases.items() if d["account"] == account}
    if engineer is not None:
        # get case number and assignee from cards so that we can determine which cases
        # belong to the engineer
        logging.warning("filtering cases for {}".format(engineer))
        cards = {
            c: d for (c, d) in cards.items() if d["assignee"]["displayName"] == engineer
        }
        temp_cases = {}
        for case, details in cases.items():
            for card in cards:
                if (
                    case == cards[card]["case_number"]
                    and cards[card]["assignee"]["displayName"] == engineer
                ):
                    temp_cases[case] = details
        cases = temp_cases
```

in any case, this might be a starting point for a solution.

**Review by adhil0 — CHANGES_REQUESTED** [review]
Thanks

File: dashboard/src/t5gweb/libtelco5g.py (hunk 5 above)
Comment (dcritch) [reply dashboard/src/t5gweb/libtelco5g.py:429]: > 
> I've been trying this, but the numbers aren't adding up (the number of "Open Cases" in the table does not match the open cards in the table below):
> 
> 
> in any case, this might be a starting point for a solution.

The tables includes both open and closed cards/cases though. Could that be the reason for the mismatch you're seeing?

Comment (adhil0) [reply dashboard/src/t5gweb/libtelco5g.py:429]: My numbers aren't in sync with production, but the logic should be the same: I'm seeing 85 cases in the table, 20 of which are closed. So under Open cases we should see 85-20 =65, but I'm seeing 44 instead

Comment (adhil0) [reply dashboard/src/t5gweb/libtelco5g.py:429]: @dcritch OK, I think I narrowed down the issue: 
1. I think the numbers from my earlier comment are off, so after refreshing:
2. The stats tables are populated using `generate_stats()`, while the interactive tables at the bottom are generated by `get_new_comments()`. `get_new_comments()` throws out cards that don't have any comments on them, while `generate_stats()` doesn't, which results in different numbers between the top and bottom halves of the page.

In production, it's quite rare for a card to have no comments, so the mismatch doesn't really show up anywhere. I think we implemented this for the reports (`<URL>updates/all`) so that every bullet would have some content, and then just plugged it into the tables when we were creating the table view. Should we remove the behavior of dropping cards with 0 comments to make everything synced up with each other?

TL;DR I believe the mismatch is outside of the scope of this PR. @ptrnull When you get a chance, could you update your branch with the suggestion mentioned above (repeated below as well for convenience):

libtelco5g.py:425 =>
```python
    if account is not None:
        logging.warning("filtering cases for {}".format(account))
        cards = {c: d for (c, d) in cards.items() if d["account"] == account}
        cases = {c: d for (c, d) in cases.items() if d["account"] == account}
    if engineer is not None:
        logging.warning("filtering cases for {}".format(engineer))
        cards = {
            c: d for (c, d) in cards.items() if d["assignee"]["displayName"] == engineer
        }

        # get case number and assignee from cards so that we can determine which cases
        # belong to the engineer
        temp_cases = {}
        for case, details in cases.items():
            for card in cards:
                if (
                    case == cards[card]["case_number"]
                    and cards[card]["assignee"]["displayName"] == engineer
                ):
                    temp_cases[case] = details
        cases = temp_cases
```

Comment (dcritch) [reply dashboard/src/t5gweb/libtelco5g.py]: great detective work! yes, maybe we should return 'No Comments' or something like that rather than drop them. we can discuss separate from this PR

**Comment by dcritch** [comment]
the 1 lint issue it is complaining about isn't a big deal in my opinion. however if you wanted to make a new function in utils.py so:
```
pie_stats = make_pie_dict(stats)
```
and in utils.py:
```
def make_pie_dict(stats):
    return {
        "by_severity": (
            list(stats["by_severity"].keys()),
            list(stats["by_severity"].values()),
        ),
        "by_status": (
            list(stats["by_status"].keys()),
            list(stats["by_status"].values()),
        ),
    }
```
that'd be swell. then update both `get_engineer` and `get_account` to use it.

**Comment by adhil0** [comment]
> the 1 lint issue it is complaining about isn't a big deal in my opinion. however if you wanted to make a new function in utils.py so:
> 
> ```
> pie_stats = make_pie_dict(stats)
> ```
> 
> and in utils.py:
> 
> ```
> def make_pie_dict(stats):
>     return {
>         "by_severity": (
>             list(stats["by_severity"].keys()),
>             list(stats["by_severity"].values()),
>         ),
>         "by_status": (
>             list(stats["by_status"].keys()),
>             list(stats["by_status"].values()),
>         ),
>     }
> ```
> 
> that'd be swell. then update both `get_engineer` and `get_account` to use it.

+1 - on second thought, it's probably better  to put this in a function since we're using the code in multiple places

File: dashboard/src/t5gweb/libtelco5g.py [hunk 6]
```diff
@@ -654,6 +673,12 @@ def generate_histogram_stats(account=None):
         logging.warning(f"filtering cards for {account}")
         cards = {c: d for (c, d) in cards.items() if d["account"] == account}
 
+    if engineer is not None:
+        logging.warning(f"filtering cards for {account}")
```
Comment (dcritch) [inline dashboard/src/t5gweb/libtelco5g.py:677]: s/account/engineer

**Review by dcritch — COMMENTED** [review]
one minor nit, then you're good to go. thanks!

File: dashboard/src/t5gweb/libtelco5g.py (hunk 6 above)
Comment (ptrnull) [reply dashboard/src/t5gweb/libtelco5g.py]: done

**Review by dcritch — APPROVED** [review]
thanks for the contribution and for dealing with all the back and forth :) lgtm!

**Review by adhil0 — APPROVED** [review]
Thanks, looking really good now
